### PR TITLE
cgroups: apply StartupAllowedCPUs= and StartupAllowedMemoryNodes= during shutdown

### LINKED ISSUE #20733: RFE: enable control of cgroup attributes at runtime
**Is your feature request related to a problem? Please describe.**

We have a system where fast re(boots) are a requirement, thus I want to be able to give all available resources to the system.slice  during boot and shutdown. However after boot (after some arbitrary target is reached) I would like to ensure processes in the system.slice cgroup only run on a subset of available CPUs (to avoid interrupting another latency sensitive application in another slice). We can use the AllowedCPUs option to pin a cgroup to a cpuset, however this requires either of:

   * AllowedCPUs is set for the system slice during boot (this slows our boot down so isn't an option).
   * A runtime drop-in is created to change AllowedCPUs at runtime. This seems like a heavy weight solution and possibly difficult to manage by another user defined service.
   * We don't set AllowedCPUs for the system.slice, rather after boot we manually twiddle the cpuset.cpus file for the system.slice. This violates the principle of a "single writer" that systemd preaches, sadly however this is the solution I am currently using.

**Describe the solution you'd like**

A less heavy weight way (as opposed to runtime drop-in creation and a daemon-reload) of changing resource configuration of a service/slice/scope. These properties are currently dbus RO, if they could be made RW then this would be a lighter weight way to change configuration. I'd be happy to work on a patch for this.

**Describe alternatives you've considered**

Copy/paste from above:

   * AllowedCPUs is set for the system slice during boot (this slows our boot down so isn't an option).
   * A runtime drop-in is created to change AllowedCPUs at runtime. This seems like a heavy weight solution and possibly difficult to manage by another user defined service.
   * We don't set AllowedCPUs for the system.slice, rather after boot we manually twiddle the cpuset.cpus file for the system.slice. This violates the principle of a "single writer" that systemd preaches, sadly however this is the solution I am currently using.

**The systemd version you checked that didn't have the feature you are asking for**

v249.

### PR DESCRIPTION
**Pull request by pdmorrow**
Allow StartupAllowedCPUs= and StartupAllowedMemoryNodes= during shutdown. This is useful for speeding up shutdown/reboot time if some more retrictive CPU affinity policy was applied during normal runtime via AllowedCPUs= and/or AllowedMemoryNodes=.

Further fixes #20733

### CODE REVIEW

File: src/core/cgroup.c [hunk 1]
```diff
@@ -389,6 +391,7 @@ static char *format_cgroup_memory_limit_comparison(char *buf, size_t l, Unit *u,
 
 void cgroup_context_dump(Unit *u, FILE* f, const char *prefix) {
         _cleanup_free_ char *disable_controllers_str = NULL, *cpuset_cpus = NULL, *cpuset_mems = NULL, *startup_cpuset_cpus = NULL, *startup_cpuset_mems = NULL;
+        _cleanup_free_ char *shutdown_cpuset_cpus = NULL, *shutdown_cpuset_mems = NULL;
```
Comment (bluca) [inline src/core/cgroup.c]: break the above line, no need to duplicate

Comment (pdmorrow) [reply src/core/cgroup.c]: Done, though unsure what the coding convention is on line splitting & indentation like this. PTAL :)

File: src/core/cgroup.c [hunk 2]
```diff
@@ -388,7 +390,8 @@ static char *format_cgroup_memory_limit_comparison(char *buf, size_t l, Unit *u,
 }
 
 void cgroup_context_dump(Unit *u, FILE* f, const char *prefix) {
-        _cleanup_free_ char *disable_controllers_str = NULL, *cpuset_cpus = NULL, *cpuset_mems = NULL, *startup_cpuset_cpus = NULL, *startup_cpuset_mems = NULL;
+        _cleanup_free_ char *disable_controllers_str = NULL, *cpuset_cpus = NULL, *cpuset_mems = NULL, \
```
Comment (bluca) [inline src/core/cgroup.c]: `\`? Been doing too much python? :-P indent the next line so that the first `*` matches

Comment (pdmorrow) [reply src/core/cgroup.c]: Yeah, let's blame it on that :) Fixed.

**Review by bluca — APPROVED** [review]
lgtm

**Comment by poettering** [comment]
Hmm, does this really matter? what kind of workloads do you have to optimize shutdown like that?

**Comment by bluca** [comment]
> Hmm, does this really matter? what kind of workloads do you have to optimize shutdown like that?

Sadly it does make a difference

**Comment by pdmorrow** [comment]
> Hmm, does this really matter? what kind of workloads do you have to optimize shutdown like that?

It matters for our sorts of workloads. We need to do quite a considerable amount of work during shutdown unfortunately, we also need CPU affinity configuration during runtime for some latency sensitive applications. Sticking with the runtime CPU affinity configuration during shutdown impacts our shutdown time quite significantly. One thing we are aiming for is very quick reboots, so anything that gets in the way of that (such as only being able to use half the systems CPUs during shutdown) is something that I would like to avoid if at all possible.

**Comment by pdmorrow** [comment]
CI failure looks unrelated?

```
../src/shared/creds-util.c: In function â€˜sha256_hash_host_and_tpm2_keyâ€™:
../src/shared/creds-util.c:412:9: error: â€˜SHA256_Initâ€™ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  412 |         if (SHA256_Init(&sha256_context) != 1)
      |         ^~
In file included from /usr/include/openssl/x509.h:41,
                 from ../src/shared/openssl-util.h:8,
                 from ../src/shared/creds-util.c:21:
/usr/include/openssl/sha.h:73:27: note: declared here
   73 | OSSL_DEPRECATEDIN_3_0 int SHA256_Init(SHA256_CTX *c);
      |                           ^~~~~~~~~~~
../src/shared/creds-util.c:415:9: error: â€˜SHA256_Updateâ€™ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  415 |         if (host_key && SHA256_Update(&sha256_context, host_key, host_key_size) != 1)
      |         ^~
In file included from /usr/include/openssl/x509.h:41,
                 from ../src/shared/openssl-util.h:8,
                 from ../src/shared/creds-util.c:21:
/usr/include/openssl/sha.h:74:27: note: declared here
   74 | OSSL_DEPRECATEDIN_3_0 int SHA256_Update(SHA256_CTX *c,
      |                           ^~~~~~~~~~~~~
../src/shared/creds-util.c:418:9: error: â€˜SHA256_Updateâ€™ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  418 |         if (tpm2_key && SHA256_Update(&sha256_context, tpm2_key, tpm2_key_size) != 1)
      |         ^~
In file included from /usr/include/openssl/x509.h:41,
                 from ../src/shared/openssl-util.h:8,
                 from ../src/shared/creds-util.c:21:
/usr/include/openssl/sha.h:74:27: note: declared here
   74 | OSSL_DEPRECATEDIN_3_0 int SHA256_Update(SHA256_CTX *c,
      |                           ^~~~~~~~~~~~~
../src/shared/creds-util.c:421:9: error: â€˜SHA256_Finalâ€™ is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  421 |         if (SHA256_Final(ret, &sha256_context) != 1)
      |         ^~
In file included from /usr/include/openssl/x509.h:41,
                 from ../src/shared/openssl-util.h:8,
                 from ../src/shared/creds-util.c:21:
/usr/include/openssl/sha.h:76:27: note: declared here
   76 | OSSL_DEPRECATEDIN_3_0 int SHA256_Final(unsigned char *md, SHA256_CTX *c);
      |                           ^~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by bluca** [comment]
yes ignore that

**Comment by poettering** [comment]
Hmm, so then let me ask a differen question:

should this really be a separate config option? wouldn't be a more appropriate fix to just start saying that the StartupXYZ= settings should also apply to shutdown? (even if they are a bit of a misnomer then?)

**Comment by bluca** [comment]
That would work for our case as it's symmetric, although it would be a practical unexpected change for others I think, no?

**Comment by poettering** [comment]
yeah, it would be strictly speaking a compat break. but i#d say a really really weak one. and i can't come up with usecase where this shouldn't be OK?

**Comment by pdmorrow** [comment]
> Hmm, so then let me ask a differen question:
> 
> should this really be a separate config option? wouldn't be a more appropriate fix to just start saying that the StartupXYZ= settings should also apply to shutdown? (even if they are a bit of a misnomer then?)

That works too, I had considered this but figured being explicit is better. However we are somewhat bloating the number of config options which is never good. I can't think of cases were you would need 3 different configuration, at least in our case we only need 2 (runtime and startup/shutdown). Maybe someone else might though...?

**Comment by bluca** [comment]
It's such a minor thing and unlikely to break something, it would most likely be fine. If someone complains then we can revisit, as it's easy enough.

**Comment by pdmorrow** [comment]
> It's such a minor thing and unlikely to break something, it would most likely be fine. If someone complains then we can revisit, as it's easy enough.

OK, sounds like I should rework this to get rid of ShutdownAllowedCPUs and ShutdownAllowedMemoryNodes and fix StartupAllowedCPUs and StartupAllowedMemoryNodes to apply to both boot and shutdown?

StartupAllowedCPUs and StartupAllowedMemoryNodes was only merged on Tuesday, so we do have an opportunity to re-name them to something better. Though I am stumped at a decent name for something that applies to Shutdown and Startup :)

**Comment by poettering** [comment]
Well, the cpu shares settings are already called "Startup", even though this would be a misnomer now. but i think this is a detail we should be able to cover for with a documentation fix, and then ignore.

**Comment by bluca** [comment]
We could maybe introduce StartupShutdown* aliases for everything and do the usual deprecation process for the old ones?

**Comment by poettering** [comment]
well, not sure its worth it. also do you have a good idea even for a word to use for the combination of "startup" and "shutdown" phases?

**Comment by poettering** [comment]
i mean we have "minor" misnomers all over the place. for example .socket units can also encapsulate message queues and other stuff. which aren't sockets. but also: meh, i don't care... ;-)

**Comment by pdmorrow** [comment]
> i mean we have "minor" misnomers all over the place. for example .socket units can also encapsulate message queues and other stuff. which aren't sockets. but also: meh, i don't care... ;-)

OK, will ditch the Shutdown directives :)

**Comment by pdmorrow** [comment]
OK, got rid of the ShutdownAllowed* directives and made StartupAllowed* also apply during shutdown - looks to work fine. Whilst I am here should I switch the other Startup* directives to do the same thing (i.e. apply during shutdown)?

**Comment by poettering** [comment]
Also, there needs to be a well placed manager_invalidate_startup_units() call now that is called when we enter startup phase.

**Comment by pdmorrow** [comment]
> Also, there needs to be a well placed manager_invalidate_startup_units() call now that is called when we enter shutdown phase.

Added this now. The reason I didn't spot this testing first time around is that I tested with slice units and not service units, the slice units are realized already during the shutdown process.

It might make sense to change the behaviour of the other Startup directives now to also apply at shutdown now so that all Startup* directives behave like this and not just StartupAllowedCPUs and StartupAllowedMemoryNodes. Happy to do that here, or via another PR :)

File: src/core/cgroup.c [hunk 3]
```diff
@@ -2543,6 +2562,12 @@ unsigned manager_dispatch_cgroup_realize_queue(Manager *m) {
         assert(m);
 
         state = manager_state(m);
+        if (state == MANAGER_STOPPING && !m->shutdown_units_refreshed) {
+                /* Units which are refreshed after startup are also refreshed
+                 * at shutdown time. */
+                manager_invalidate_startup_units(m);
+                m->shutdown_units_refreshed = true;
+        }
```
Comment (poettering) [inline src/core/cgroup.c]: this doesn't look like the right place to insert this.

if you look into manager_state() you'll find that MANAGER_STOPPING is returned whenever the special unit "shutdown.target" is enqueued or active. Thus, it's important to invalidate the startup unit cgroups stuff whenever that unit is enqueued. i.e. add a call to manager_invalidate_startup_units() to manager_add_job() if the right unit is enqueued.

Comment (pdmorrow) [reply src/core/cgroup.c]: I don't think SPECIAL_SHUTDOWN_TARGET is directly enqueued as a job? I see SPECIAL_{HALT,POWEROFF,REBOOT,KEXEC}_TARGET be enqueued but not SPECIAL_SHUTDOWN_TARGET. The later seems to be installed indirectly.

Thus I don't think I can add a general catch all in manager_add_job() which would call manager_invalidate_startup_units() when any of the HALT/POWEROFF/REBOOT/KEXEC jobs are added. I could write some code to catch all these targets but that doesn't seem very nice.

I've pushed changes which do work, manager_invalidate_startup_units() is called via job_install(). This doesn't feel great either, so looking for your feedback again :)

Comment (pdmorrow) [reply src/core/cgroup.c]: @poettering would you be able to take a look when you have a moment please?  Thanks!

File: src/core/job.c [hunk 4]
```diff
@@ -239,6 +239,9 @@ Job* job_install(Job *j) {
                        "Installed new job %s/%s as %u",
                        j->unit->id, job_type_to_string(j->type), (unsigned) j->id);
 
+        if (j->unit == manager_get_unit(j->manager, SPECIAL_SHUTDOWN_TARGET))
+                manager_invalidate_startup_units(j->manager);
```
Comment (poettering) [inline src/core/job.c]: move to job_shutdown_magic()

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/core/job.c (hunk 4 above)
Comment (pdmorrow) [reply src/core/job.c]: Fixed & re-tested. Thanks.

File: src/core/job.c [hunk 5]
```diff
@@ -1338,6 +1338,8 @@ void job_shutdown_magic(Job *j) {
         if (detect_container() > 0)
                 return;
 
+        manager_invalidate_startup_units(j->unit->manager);
```
Comment (poettering) [inline src/core/job.c]: i see no reason to conditionalize this on detect_container(). if we have the perms to muck with cgroups, then i think we should do this in a container too if configured that way.

i.e. just move this line before the detect_container() check

Comment (pdmorrow) [reply src/core/job.c]: Yeah, absolutely. I missed this...

**Comment by poettering** [comment]
so patch looks pretty good now. one more thing though: I'd really prefer if all StartupXYZ= settings would behave the same here, i.e. StartupCPUWeight= and StartupIOWeight= should behave the same way here really and also be applied on shutdown again

**Comment by pdmorrow** [comment]
> so patch looks pretty good now. one more thing though: I'd really prefer if all StartupXYZ= settings would behave the same here, i.e. StartupCPUWeight= and StartupIOWeight= should behave the same way here really and also be applied on shutdown again

Yep, that's fair. I'll make the change.

**Comment by poettering** [comment]
excellent! thanks! lgtm!

**Comment by pdmorrow** [comment]
> excellent! thanks! lgtm!

Thanks for the review & your time!

**Pull request merged**